node/src/components/linear_chain.rs

```diff
@@ -98,29 +100,75 @@ impl<I: Display> Display for Event<I> {
                 height,
                 block.is_some()
             ),
+            Event::GetBlockForFinalitySignaturesResult(block_hash, maybe_block) => {
+                write!(
+                    f,
+                    "linear chain get-block-for-finality-signatures-result for {} found: {}",
+                    block_hash,
+                    maybe_block.is_some()
+                )
+            }
         }
     }
 }
 
 #[derive(DataSize, Debug)]
 pub(crate) struct LinearChain<I> {
-    /// A temporary workaround.
-    // TODO: Refactor to proper LRU cache.
-    linear_chain: Vec<Block>,
+    /// The most recently added block.
+    latest_block: Option<Block>,
+    /// Finality signatures to be inserted in a block once it is available.
+    pending_finality_signatures: HashMap<PublicKey, HashMap<BlockHash, FinalitySignature>>,
     _marker: PhantomData<I>,
 }
 
 impl<I> LinearChain<I> {
     pub fn new() -> Self {
         LinearChain {
-            linear_chain: Vec::new(),
+            latest_block: None,
+            pending_finality_signatures: HashMap::new(),
             _marker: PhantomData,
         }
     }
 
     // TODO: Remove once we can return all linear chain blocks from persistent storage.
-    pub fn linear_chain(&self) -> &Vec<Block> {
-        &self.linear_chain
+    pub fn latest_block(&self) -> &Option<Block> {
+        &self.latest_block
+    }
+
+    /// Adds pending finality signatures to the block; returns events to announce and broadcast
+    /// them.
+    fn add_pending_finality_signatures<REv>(
+        &mut self,
+        block: &mut Block,
+        effect_builder: EffectBuilder<REv>,
+    ) -> Effects<Event<I>>
+    where
+        REv: From<StorageRequest>
+            + From<ConsensusRequest>
+            + From<NetworkRequest<I, Message>>
+            + From<LinearChainAnnouncement>
+            + Send,
+        I: Display + Send + 'static,
+    {
+        let mut effects = Effects::new();
+        let block_hash = block.hash();
+        let pending_sigs = self
+            .pending_finality_signatures
+            .values_mut()
+            .filter_map(|sigs| sigs.remove(&block_hash).map(Box::new))
+            // TODO: Store signatures by public key.
+            .filter(|fs| !block.proofs().iter().any(|proof| proof == &fs.signature))
+            .collect_vec();
+        self.pending_finality_signatures
+            .retain(|_, sigs| !sigs.is_empty());
+        // Add new signatures and send the updated block to storage.
+        for fs in pending_sigs {
+            block.append_proof(fs.signature);
+            let message = Message::FinalitySignature(fs.clone());
+            effects.extend(effect_builder.broadcast_message(message).ignore());
+            effects.extend(effect_builder.announce_finality_signature(fs).ignore());
+        }
+        effects
```

Comment:
We should store the block again with the new signatures.

Reply:
Ah, it is stored but the means of mutating the argument `block` :grimacing: please, don't do that. That's horrible. Let's return `(Block, Effects<Event<I>>)`.

Reply:
Done in https://github.com/CasperLabs/casper-node/pull/649/commits/8404eced79d3716e3393a238fa2d18ad6a60db17.